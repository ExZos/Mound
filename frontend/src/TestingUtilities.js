import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'user' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
