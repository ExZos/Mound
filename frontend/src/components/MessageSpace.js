import React from 'react';
import { TextField, CircularProgress } from '@material-ui/core';

import GeneralComponent from './GeneralComponent';
import PollSpace from './PollSpace';
import NamePoll from './NamePoll';
import { server, api } from '../server';
import '../styles/messageSpace.css';

class MessageSpace extends GeneralComponent {
  constructor(props) {
    super(props);

    this.loaded = false;

    this.state = {
      messages: [],
      message: {
        user: this.props.user.id,
        content: ''
      }
    };
  }

  componentDidMount() {
    // TEMP
    console.log(this.props.user.id + " : " + this.props.user.name + " : " + this.props.user.space);

    this.getMessages();
    this.interval = setInterval(this.getMessages, 1000);
  }

  componentDidUpdate() {
    window.onpopstate = (e) => {
      this.loaded = false;

      this.props.updateState();
      this.getMessages();
    }
  }

  componentWillUnmount() {
    this.loaded = false;

    clearInterval(this.interval);
  }

  getMessages = async () => {
    try {
      const res = await server.get(api.getMessagesInSpace + this.props.user.space);

      this.loaded = true;

      this.setState({
        messages: res.data
      });

      this.updateUserLastActive();
    } catch (e) {
      // TODO: render error component
    }
  }

  updateUserLastActive = async () => {
    try {
      const res = await server.put(api.users + this.props.user.id + '/', this.props.user);

      if(this.loaded) {
        this.addToSessionArrayItem('users', res.data);

        this.props.updateState();
      }
    } catch (e) {
      // User deleted: force logout
      // TODO: implement bans (User.banned)
      this.removeSessionArrayItem('users', this.props.user.space);

      this.props.history.push({
        pathname: '/'
      });
    }
  }

  addMessage = async () => {
    try {
      /*const res = */await server.post(api.messages, this.state.message);
      this.getMessages();

      this.setState({
        message: {
          user: this.props.user.id,
          content: ''
        }
      });
    } catch (e) {
      // TODO: render error component
    }
  }

  handleKeyFormSubmit = (e) => {
    if(e.keyCode === 13 && e.shiftKey === false) {
      e.preventDefault();
      this.addMessage();
    }
  }

  toggleMessageTimestamp = (e) => {
    // console.log(e.target.className);
    if(e.target.nextSibling.style.display) {
      e.target.nextSibling.style.display = "";
    }
    else {
      e.target.nextSibling.style.display = "inherit";
    }
  }

  renderMessages = () => {
    const messages = this.state.messages;

    // No messages
    if(messages.length === 0) {
      return (
        <div className="noMessages">
          Not much of a convo here huh?
        </div>
      );
    }

    return messages.map(message => (
      <div key={message.id}
        className={(message.user === this.props.user.id) ? "message own" : "message"}
      >
        <div className="sender">
          {(message.user === this.props.user.id) ? "You" : message.user_name}
        </div>

        <div className="content" onClick={this.toggleMessageTimestamp}>
          {message.content}
        </div>

        <div className="timestamp">
          {this.convertTimestamp(message.timestamp)}
        </div>
      </div>
    ));
  }

  render() {
    if(!this.loaded) {
      return (
        <div id="messageSpace">
          <CircularProgress color="inherit" />
        </div>
      );
    }

    return(
      <div id="messageSpace">
        <div className="messages">
          {this.renderMessages()}
        </div>

        <div>
          <form id="createMessage" onSubmit={this.handleFormSubmit}>
            <TextField name="content" placeholder="Type a message... " autoFocus
              variant="outlined" multiline fullWidth
              value={this.state.message.content}
              onChange={(e) => this.handleInputChange(e, 'message')}
              onKeyDown={this.handleKeyFormSubmit}
            />
          </form>
        </div>

        <br />

        <PollSpace spaceID={this.props.user.space} userID={this.props.user.id} />

        <br />

        <NamePoll spaceID={this.props.user.space} userID={this.props.user.id} />
      </div>
    );
  }
}

export default MessageSpace;
