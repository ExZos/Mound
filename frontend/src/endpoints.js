export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
