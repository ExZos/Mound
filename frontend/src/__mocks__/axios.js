import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
