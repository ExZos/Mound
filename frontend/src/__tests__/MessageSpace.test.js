import React from 'react';
import { shallow, mount } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';
import styled from 'styled-components';
import 'jest-styled-components';

import MessageSpace from '../components/MessageSpace';
import TestingUtilities from '../TestingUtilities';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('MessageSpace', () => {
  beforeEach(() => {
    tutils.addUserSessionItem(1, 1);
    tutils.addUserSessionItem(2, 2);
    tutils.addUserSessionItem(3, 3);
  });

  it('should render spinner while loading messages with props and session', () => {
    const component = shallow(<MessageSpace spaceID={1} updateState={jest.fn()} history={history} />);
    const messages = component.find('div.message');
    const spinner = component.find('Spinner');

    expect(component).toMatchSnapshot();
    expect(messages).toHaveLength(0);
    expect(spinner).toHaveLength(1);
  });

  it('should render no messages correctly with props and session', async () => {
    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<MessageSpace spaceID={1} updateState={jest.fn()} history={history} />);
    const messages = component.find('div.message');
    const noMessages = component.find('div.noMessages');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getMessagesInSpace + 1);
    expect(messages).toHaveLength(0);
    expect(noMessages).toHaveLength(1);
  });

  it('should render messages correctly with props and session', async () => {
    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<MessageSpace spaceID={2} updateState={jest.fn()} history={history} />);
    const messages = component.find('div.message');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getMessagesInSpace + 2);
    expect(messages).toHaveLength(5);
  });

  it('should mark own messages correctly', async () => {
    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<MessageSpace spaceID={2} updateState={jest.fn()} history={history} />);
    const messages = component.find('div.message.own div.sender');

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getMessagesInSpace + 2);
    expect(messages).toHaveLength(2);
    expect(messages.at(0).text()).toEqual('You');
    expect(messages.at(1).text()).toEqual('You');
  });

  it('should create a state entry equal to the input value', async () => {
    const component = await shallow(<MessageSpace spaceID={2} updateState={jest.fn()} history={history} />);
    const textarea = component.find('textarea');

    textarea.props().onChange({
      target: {
        name: 'content',
        value: 'content1'
      }
    });

    expect(component.state('message')).toBeDefined();
    expect(component.state('message')).toHaveProperty('content', 'content1');
  });

  it('should call post api on message submit', async () => {
    const component = await shallow(<MessageSpace spaceID={2} updateState={jest.fn()} history={history} />);
    const textarea = component.find('textarea');
    const spy = jest.spyOn(axios, 'post');
    const mockPreventDefault = jest.fn();

    component.setState({
      message: {
        id: 2,
        content: 'content1'
      }
    });
    textarea.props().onKeyDown({
      keyCode: 13,
      shiftKey: false,
      preventDefault: mockPreventDefault
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.messages, {
      id: 2,
      content: 'content1'
    });
  });

  // TODO: finish this
  it('should show message timestamp on message content click', async () => {
    // const component = await shallow(<MessageSpace spaceID={2} updateState={jest.fn()} history={history} />);
    // const messageContents = component.find('div.content');
    // const messageTimestamps = component.find('div.timestamp');

    // expect(messageTimestamps.at(0)).toHaveStyleRule('display', 'inherit');
  });
});
