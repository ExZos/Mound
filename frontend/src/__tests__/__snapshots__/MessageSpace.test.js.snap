// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageSpace should render CircularProgress while loading messages with props and session 1`] = `
<div
  id="messageSpace"
>
  <WithStyles(ForwardRef(CircularProgress))
    color="inherit"
  />
</div>
`;

exports[`MessageSpace should render messages correctly with props and session 1`] = `
<div
  id="messageSpace"
>
  <div
    className="messages"
  >
    <div
      className="message"
      key="1"
    >
      <div
        className="sender"
      />
      <div
        className="content"
        onClick={[Function]}
      >
        First!
      </div>
      <div
        className="timestamp"
      >
        1969-12-31, 7:00:00 p.m.
      </div>
    </div>
    <div
      className="message own"
      key="2"
    >
      <div
        className="sender"
      >
        You
      </div>
      <div
        className="content"
        onClick={[Function]}
      >
        Second!
      </div>
      <div
        className="timestamp"
      >
        1969-12-31, 7:00:00 p.m.
      </div>
    </div>
    <div
      className="message"
      key="3"
    >
      <div
        className="sender"
      />
      <div
        className="content"
        onClick={[Function]}
      >
        Third!
      </div>
      <div
        className="timestamp"
      >
        1969-12-31, 7:00:00 p.m.
      </div>
    </div>
    <div
      className="message own"
      key="4"
    >
      <div
        className="sender"
      >
        You
      </div>
      <div
        className="content"
        onClick={[Function]}
      >
        Fourth!
      </div>
      <div
        className="timestamp"
      >
        1969-12-31, 7:00:00 p.m.
      </div>
    </div>
    <div
      className="message"
      key="5"
    >
      <div
        className="sender"
      />
      <div
        className="content"
        onClick={[Function]}
      >
        Fifth!
      </div>
      <div
        className="timestamp"
      >
        1969-12-31, 7:00:00 p.m.
      </div>
    </div>
  </div>
  <div>
    <form
      id="createMessage"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        fullWidth={true}
        multiline={true}
        name="content"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Type a message... "
        value=""
        variant="outlined"
      />
    </form>
  </div>
  <br />
  <PollSpace
    spaceID={2}
    userID={2}
  />
  <br />
  <NamePoll
    spaceID={2}
    userID={2}
  />
</div>
`;

exports[`MessageSpace should render no messages correctly with props and session 1`] = `
<div
  id="messageSpace"
>
  <div
    className="messages"
  >
    <div
      className="noMessages"
    >
      Not much of a convo here huh?
    </div>
  </div>
  <div>
    <form
      id="createMessage"
      onSubmit={[Function]}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        fullWidth={true}
        multiline={true}
        name="content"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Type a message... "
        value=""
        variant="outlined"
      />
    </form>
  </div>
  <br />
  <PollSpace
    spaceID={1}
    userID={1}
  />
  <br />
  <NamePoll
    spaceID={1}
    userID={1}
  />
</div>
`;
